Adding test-federation-cmd.sh to test kubectl with federation apiserver

nikhiljindal opened the pull request:
There are 2 parts to the PR:
* Adding ability to run kubectl tests for a subset of resources.
* Adding test-federation-cmd.sh that runs kubectl tests for resources that are supported by federation-apiserver.

cc @kubernetes/sig-federation @kubernetes/sig-api-machinery 

```release-note
Adding kubectl tests for federation
```

nikhiljindal committed:
Adding test-federation-cmd.sh to test kubectl with federation apiserver

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/38844)
<!-- Reviewable:end -->

nikhiljindal commented:
Fixes https://github.com/kubernetes/kubernetes/issues/32826
Supersedes https://github.com/kubernetes/kubernetes/pull/33414

nikhiljindal commented:
Rebased. Unfortunately I will have to rebase each time someone updates hack/make-rules/test-cmd.sh

madhusudancs commented on hack/testdata/kubernetes-service.yaml:
```diff
@@ -0,0 +1,16 @@
+    protocol: TCP
+    targetPort: 443
+  sessionAffinity: ClientIP
+  type: ClusterIP
```
The above two fields are defaults right? Do you need them?

nikhiljindal replied:
removed

madhusudancs commented on hack/testdata/kubernetes-service.yaml:
```diff
@@ -0,0 +1,16 @@
+    targetPort: 443
+  sessionAffinity: ClientIP
+  type: ClusterIP
+status:
```
Why this field in the yaml?

nikhiljindal replied:
removed

madhusudancs reviewed (commented):
@nikhiljindal mostly minor nits.

madhusudancs reviewed (commented):
@nikhiljindal mostly minor nits.

nikhiljindal commented on hack/lib/test.sh:
```diff
@@ -23,7 +23,12 @@ readonly   red=$(tput setaf 1)
 
 kube::test::clear_all() {
-  kubectl delete "${kube_flags[@]}" rc,pods --all --grace-period=0 --force
+  if kube::test::if_supports_resource "rc" ; then
```
Following the convention of other kube::test functions. Ex: kube::test::if_has_string

madhusudancs replied:
Ack.

janetkuo replied:
Use variables here as well?

nikhiljindal commented on hack/lib/test.sh:
```diff
@@ -223,3 +228,20 @@ kube::test::if_has_string() {
+  REQUIRED_RESOURCE=${1:-""}
+
+  for r in "${SUPPORTED_RESOURCES[@]}"; do
+    if [[ "${r}" == "*" || "${r}" == "${REQUIRED_RESOURCE}" ]]; then
```
its expected that SUPPORTED_RESOURCE will be an explicit list of supported resources or just "*" which is the default value which means that all resources are supported.

madhusudancs replied:
Yeah, but is that expectation asserted somewhere? What's the defined behavior when the input is my example above?

nikhiljindal replied:
The expected behavior is that if supported resources array contains "*" then it supports all resources. In your example, if SUPPORTED_RESOURCES=(rc rs *), then the function will return true for all resources.

madhusudancs replied:
Ok fair enough. Please document that somewhere, with examples. Documenting it in function docstring is fine.

nikhiljindal replied:
Done. Updated the documentation for the method.

nikhiljindal commented on hack/make-rules/test-federation-cmd.sh:
```diff
@@ -0,0 +1,82 @@
+#!/bin/bash
+
+# Copyright 2014 The Kubernetes Authors.
```
fixed

nikhiljindal commented on hack/lib/test.sh:
```diff
@@ -223,3 +228,20 @@ kube::test::if_has_string() {
+
+  for r in "${SUPPORTED_RESOURCES[@]}"; do
+    if [[ "${r}" == "*" || "${r}" == "${REQUIRED_RESOURCE}" ]]; then
+      return 0
```
Same. following the convention of other methods here.
This is better than returning true or false since then the check will have to be explicit
`if [[ if_supports_resource "rc" == "true" ]]; then` `if [[ if_supports_resource "rc" ]]; then` will not work.
http://stackoverflow.com/questions/5431909/bash-functions-return-boolean-to-be-used-in-if

madhusudancs replied:
Ignore this. I am not sure what I was thinking. Bash doesn't allow returning non-numeric exit codes.

nikhiljindal commented on hack/lib/test.sh:
```diff
@@ -223,3 +228,20 @@ kube::test::if_has_string() {
+      return 0
+    fi
+  done
+  return 1
```
same as above

nikhiljindal commented:
Thanks for the review @madhusudancs 
Pushed a new version as per comments.
PTAL

janetkuo commented:
Is it possible to separate "just moving codes" and "real code changes" for changes in `hack/make-rules/test-cmd.sh` and `hack/make-rules/test-cmd-util.sh`? It'll be easier to review if they're separate commits

nikhiljindal commented:
> Is it possible to separate "just moving codes" and "real code changes" for changes in hack/make-rules/test-cmd.sh and hack/make-rules/test-cmd-util.sh? It'll be easier to review if they're separate commits

@janetkuo All changes in `test-cmd.sh` and `test-cmd-util.sh` are mostly just code moves so moving them to a separate commit is not going to help. I have just split the `runTests()` method into smaller `run_<resourcename>_tests()` (like `run_pod_tests()` and `run_deployment_tests()`) and added an `if_supports_resource` condition before calling the corresponding function. Hope this helps.

nikhiljindal commented:
Not sure why the bazel test failed. Fetching logs fails. Retrying
@k8s-bot bazel test this

nikhiljindal commented:
All tests passed! :)

janetkuo commented on hack/lib/test.sh:
```diff
@@ -223,3 +228,20 @@ kube::test::if_has_string() {
+#   SUPPORTED_RESOURCES: Array of all resources supported by the apiserver. (*)
+#   means it supports all resources.
+#   $1: Name of the resource to be tested.
+kube::test::if_supports_resource() {
```
From the build log, almost every check failed (pods, rc, etc.) and thus are skipped https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/38844/pull-kubernetes-unit/10562/build-log.txt

What worries me is that `test-cmd.sh` still passed. How do we make sure this change didn't break / skip some tests by accident?

nikhiljindal replied:
Good catch. Fixed the issue and also added a check to ensure that runTests does not complete without running any tests. It does not check that all tests were run but verifies that atleast a few were run.

nikhiljindal commented on hack/make-rules/test-cmd.sh:
```diff
@@ -276,2703 +122,16 @@ else
-  kubectl config set-context "${CONTEXT}" --namespace="namespace${i}"
-}
-
-runTests() {
```
I had added it long back when we had both "v1" and "v1beta1". We always pass "v1" everywhere now so its obsolete.

nikhiljindal commented on hack/make-rules/test-cmd.sh:
```diff
@@ -276,2703 +122,16 @@ else
-  # POD creation / deletion #
-  ###########################
-
-  kube::log::status "Testing kubectl(${version}:pods)"
```
same as above

nikhiljindal commented:
Rebased and updated code as per comments.
Have pushed new changes as a new commit to make it easier to review.

PTAL.

nikhiljindal commented:
All required tests passed :)

janetkuo commented on hack/make-rules/test-cmd-util.sh:
```diff
@@ -0,0 +1,3046 @@
+  kube::test::get_object_assert rc "{{range.items}}{{$id_field}}:{{end}}" ''
+  # Command
+  ! kubectl create -f hack/testdata/recursive/rc --recursive "${kube_flags[@]}"
+  # Post-condition: frontend replication controller 
```
I'm curious why we need to create default namespace and kubernetes service now? When won't them exist?

janetkuo replied:
Also can we move it to the top or its own section, not a part of kubectl local proxy tests

janetkuo replied:
Where is this function?

janetkuo replied:
Again, test passed even though this command is not found

nikhiljindal replied:
Sorry I did not understand. This is just a comment :)

nikhiljindal replied:
We need them for federation. Federation does not have a default namespace and kubernetes service by default

janetkuo replied:
I was pointing at the line referencing `run_kubectl_request_timeout_tests`. GitHub is weird.

janetkuo replied:
This part was referencing the default namespace and kubernetes service. I'm not sure why it's off

janetkuo replied:
`kube::test::if_supports_resource "xxxxxxxx"`: it's very easy to make a typo in "xxxxxx" without any errors. Can we make it more robust?

janetkuo replied:
I was referencing the line `kube::test::if_supports_resource "horizontalpodautoscalers"`

janetkuo replied:
(Maybe because this file is too large GitHub decides to point all reviews after `L1247` to `L1247` 😞)

nikhiljindal replied:
Moved to the top

nikhiljindal replied:
Great catch Janet. Fixed. Also filed https://github.com/kubernetes/kubernetes/issues/39168 to prevent such errors to happen again

nikhiljindal replied:
Created variables for these to reduce the chances of typo

janetkuo commented on hack/make-rules/test-cmd-util.sh:
```diff
@@ -0,0 +1,3046 @@
+  kubectl delete deployment nginx "${kube_flags[@]}"
+
+  # Test kubectl create deployment
+  kubectl create deployment test-nginx --image=gcr.io/google-containers/nginx:test-cmd
```
This shouldn't belong to kubectl run testes

nikhiljindal replied:
yes moved to create_deployment_tests

janetkuo commented:
I'm basically comparing the 3000+ lines of changes line by line :( I'm still worried that this change will break some / skip some parts of changes, based on what I just found.

nikhiljindal commented:
Thanks for the review @janetkuo Updated code as per comments.
Also squashed commits

nikhiljindal commented:
@k8s-bot bazel test this

nikhiljindal commented:
There are no logs. And verify-bazel passes for me locally. Not sure whats happening here

janetkuo commented on hack/lib/test.sh:
```diff
@@ -23,7 +23,12 @@ readonly   red=$(tput setaf 1)
+  if kube::test::if_supports_resource "rc" ; then
+    kubectl delete "${kube_flags[@]}" rc --all --grace-period=0 --force
+  fi
+  if kube::test::if_supports_resource "pods" ; then
```
and here?

janetkuo commented:
LGTM. Feel free to tag once all green.

nikhiljindal commented:
Retrying @k8s-bot bazel test this
cc @k8s-oncall @mikedanese No idea why bazel is failing. It passes locally. logs are not accessible

madhusudancs commented:
LGTM from my side as well.

nikhiljindal commented:
Bazel passed :)

Adding LGTM label
